include/networkit/graph/Graph.hpp

```diff
@@ -419,11 +422,352 @@ class Graph final {
+        using iterator_category = std::forward_iterator_tag;
+
+        // Signed integer type of the result of subtracting two pointers,
+        // required by STL.
+        using difference_type = ptrdiff_t;
+
+        // Own type.
+        using self = NodeIterator;
+
+        NodeIterator(const Graph *G, node u) : G(G), u(u) {
+            if (!G->hasNode(u) && u < G->upperNodeIdBound()) {
+                ++(*this);
+            }
+        }
+
+        /**
+         * @brief WARNING: This constructor is required for Python and should not be used as the
+         * iterator is not initialized.
+         */
+        NodeIterator() : G(nullptr) {}
+
+        ~NodeIterator() = default;
+
+        NodeIterator operator++() {
+            assert(u < G->upperNodeIdBound());
+            do {
+                ++u;
+            } while (!(G->hasNode(u) || u >= G->upperNodeIdBound()));
+            return *this;
+        }
+
+        NodeIterator operator++(int) {
+            const auto tmp = *this;
+            ++(*this);
+            return tmp;
+        }
+
+        NodeIterator operator--() {
+            assert(u);
+            do {
+                --u;
+            } while (!G->hasNode(u));
+            return *this;
+        }
+
+        NodeIterator operator--(int) {
+            const auto tmp = *this;
+            --(*this);
+            return tmp;
+        }
+
+        bool operator==(const NodeIterator &rhs) const noexcept { return u == rhs.u; }
+
+        bool operator!=(const NodeIterator &rhs) const noexcept { return !(*this == rhs); }
+
+        node operator*() const noexcept {
+            assert(u < G->upperNodeIdBound());
+            return u;
+        }
+    };
+
+    /**
+     * Wrapper class to iterate over a range of the nodes of a graph.
+     */
+    class NodeRange {
+
+        const Graph *G;
+
+    public:
+        NodeRange(const Graph &G) : G(&G) {}
+
+        NodeRange() : G(nullptr){};
+
+        ~NodeRange() = default;
+
+        NodeIterator begin() const noexcept {
+            assert(G);
+            return NodeIterator(G, node{0});
+        }
+
+        NodeIterator end() const noexcept {
+            assert(G);
+            return NodeIterator(G, G->upperNodeIdBound());
+        }
+    };
+
+    // Necessary for friendship with EdgeIterator.
+    class EdgeWeightIterator;
+
+    /**
+     * Class to iterate over the edges of the graph. If the graph is undirected, operator*() returns
+     * the edges (u, v) s.t. u <= v.
+     */
+    class EdgeIterator {
+        friend class EdgeWeightIterator;
+
+        const Graph *G;
+        NodeIterator nodeIter;
+        index i;
+
+        // Whether the current edge exists, or the iteration finished.
+        bool inValidState() const noexcept {
+            return nodeIter == G->nodeRange().end()
+                   || (G->degree(*nodeIter)
+                       && (G->isDirected() || (*nodeIter < G->getIthNeighbor(*nodeIter, i))));
+        }
```

Comment:
I suggest to move out the node iterator and degree conditions out of this function (it can only be violated in nextEdge, not in prevEdge) and rename it to `wantCurrentEdge` or something similar.